README
---------------------------------------------------------------------------

    24 February 2004                  Jeffrey W. Baker <jwbaker@acm.org>    

---------------------------------------------------------------------------

DESCRIPTION
-----------

This is Apache::Session 1.93

These modules provide persistent storage for arbitrary data, in arbitrary
backing stores.  The details of interacting with the backing store are
abstracted to make all backing stores behave alike.  The programmer simply
interacts with a tied hash.

COPYRIGHT AND LICENSE
---------------------

Apache::Session is Copyright (c) 1998, 1999, 2000, 2001, 2004  Jeffrey William
Baker <jwbaker@acm.org>.  Distribute under the same terms as Perl itself.

Apache::Session is Copyright (c) 2004-2006 Casey West, <casey@geeknest.com>.
Disribute under the same terms as Perl itself.

Apache::Session is Copyright (c) 2007-2014 Alexandr Ciornii.
Disribute under the same terms as Perl itself.

PREREQUISITES
-------------

Required for testing: Test::More, Test::Deep, Test::Exception, File::Temp
Required for work: Storable, constant

Most of Apache::Session's functions require Digest::MD5.

If you wish to use the DBI backing stores, you need the latest version of DBI
and the driver for your database.  Currently MySQL, Postgres, and Oracle are
supported.

INSTALLATION
------------
Auto
Use CPAN's or CPANPLUS's shell.

Manual:


tar -xvzf Apache-Session-1.91.tar.gz
cd Apache-Session-1.91
perl Makefile.PL
make
make test
make install

Substitute "nmake"/"dmake" for "make" above if you are using Windows.

EXAMPLE
-------

See eg/example.perl.  This program is intended to be run under mod_perl.

AUTHORS
-------

Jeffrey Baker <jwbaker@acm.org> is the author of 
Apache::Session.

Tatsuhiko Miyagawa <miyagawa@bulknews.net> is the author of 
Generate::ModUniqueID and Generate::ModUsertrack

Erik Rantapaa <rantapaa@fanbuzz.com> found errors in both Lock::File
and Store::File

Bart Schaefer <schaefer@zanshin.com> notified me of a bug in 
Lock::File.

Chris Winters <cwinters@intes.net> contributed the Sybase code.

Michael Schout <mschout@gkg.net> fixed a commit policy bug in 1.51.

Andreas J. Koenig <andreas.koenig@anima.de> contributed valuable CPAN
advice and also Apache::Session::Tree and Apache::Session::Counted.

Gerald Richter <richter@ecos.de> had the idea for a tied hash interface
and provided the initial code for it.  He also uses Apache::Session in
his Embperl module and is the author of Apache::Session::Embperl

Jochen Wiedmann <joe@ipsoft.de> contributed patches for bugs and
improved performance.

Steve Shreeve <shreeve@uci.edu> squashed a bug in 0.99.0 whereby
a cleared hash or deleted key failed to set the modified bit.

Peter Kaas <Peter.Kaas@lunatech.com> sent quite a bit of feedback
with ideas for interface improvements.

Randy Harmon <rjharmon@uptimecomputers.com> contributed the original
storage-independent object interface with input from:

  Bavo De Ridder <bavo@ace.ulyssis.student.kuleuven.ac.be>
  Jules Bean <jmlb2@hermes.cam.ac.uk>
  Lincoln Stein <lstein@cshl.org>
  
Scott McWhirter <scott@surreytech.co.uk> contributed verbose error messages for
file locking.

Corris Randall <corris@line6.net> gave us the option to use any table name in
the MySQL store.

Oliver Maul <oliver.maul@ixos.de> updated the Sybase modules

Innumerable users sent a patch for the reversed file age test in the file
locking module.

Mike Langen <mike.langen@tamedia.ch> contributed Informix modules.

FURTHER INFORMATION:
--------------------

Apache   by Apache Group    comp.infosystems.www.servers.unix
                            http://www.apache.org/
                            
mod_perl by Doug MacEachern modperl@apache.org
                            http://perl.apache.org/
                            


CHANGES
In near future Apache::Session may require perl 5.6.
If you use perl older than 5.6 - write to alexchorny[AT]gmail.com
and tell me what version do you use, why and will you need newer versions
of Apache::Session.

1.94  2020-09-19
  - better error if LockDataSource is missing in Apache::Session::Lock::MySQL

1.93  2014-04-12
  - even more test fixes

1.92  2014-03-08
  - more test fixes

1.91  2014-01-07 by Alexandr Ciornii, Perl 26th birthday version
  - Add a test for RT#50896
  - 99mysql.t will work more correctly in some corner cases

1.90   2013-01-27 by Alexandr Ciornii, Perl 25th birthday version
  - Allow specifying table for Oracle
  - Use Test::Database for tests

1.89   2010-09-22  by Alexandr Ciornii
  - require Digest::MD5
  - all semaphore tests were removed
  - Apache::Session::Store::File::materialize should not append to $session->{serialized}
  - Apache::Session::Store::File will flush after writing to file

1.88   2008-12-20  by Alexandr Ciornii
  - Apache::Session::Generate::MD5::validate will untaint data
  - MIN_PERL_VERSION in Makefile.PL

1.87=1.86_03   2008-08-08
1.86_03   2008-08-03 by Alexandr Ciornii
  - disabled 99semaphore.t. Help needed.

1.86_02   2008-06-27 by Alexandr Ciornii
  - correct number of tests in 99semaphore.t

1.86_01   2008-06-20 by Alexandr Ciornii
  - use Carp::confess in Apache::Session::Lock::Semaphore
  - more info in die message in Apache::Session::Generate::MD5 (RT#36412, by Knut Arne Bjorndal)
  - semaphore tests check for semctl availability
  - Makefile.PL prints message if perl version < 5.006

1.86   2008-02-01 by Alexandr Ciornii
  - Default number of semaphores for *BSD is 8 in Apache::Session::Lock::Semaphore

1.85_01   2008-01-24 by Alexandr Ciornii
  - typos corrected (catched by Gerald Fox)
  - more tests and diag in 99semaphore.t
  - no warning "disconnect invalidates 2 active statement" in
    Apache::Session::Lock::MySQL by Tony Cook (RT#32148)

1.85   2007-12-20 by Alexandr Ciornii, Perl 20th birthday version
  - mention Catalyst::Plugin::Session, Session

1.84_01   2007-11-26 by Alexandr Ciornii (alexchorny AT gmail.com)
  - Added Apache::Session::Lock::Semaphore::remove to remove semaphore
  - 99flex.t will remove semaphore (RT#30440)
  - 99flex.t should work on 5.6.1 again (no chdir now)
  - 99flex.t will clean all temporary files (RT#30209)
  - pod.t included in MANIFEST
  - cleaner tests

1.84   2007-10-02 by Alexandr Ciornii (alexchorny AT gmail.com)
  - Added constant.pm to the list of prerequisites
  - Jeffrey W. Baker, Casey West, Alexandr Ciornii, Oliver Maul agreed to
    change license of all files to Perl license
    Alexandr Ciornii agrees to relicense to Artistic 2.0 in future if needed.
  - 99flex.t will be skipped on *bsd and Solaris

1.83_01   2007-08-03 by Alexandr Ciornii
  - better handling of Storable errors by Rick Delaney (RT#27476)

1.83=1.82_05  2007-05-25
1.82_05   2007-05-14 by Alexandr Ciornii
  - skipping part of 99flex.t on NetBSD

1.82_04   2007-04-27 by Alexandr Ciornii
  - More diagnostics in Apache::Session::Lock::Semaphore::acquire_write_lock,
    acquire_read_lock
  - Did not increment modules versions in previous versions of distribution
  - Apache::Session::Lock::Semaphore can work with private semaphore
  - 99flex.t will use private semaphore
  - 99dbfile.t, 99dbfilestore.t - added undef (for RT#6216)

1.82_03   2007-03-12 by Alexandr Ciornii
  - Apache::Session::Lock::File checks flock success (RT#6936)
  - Apache::Session::Lock::File will not change to shared lock if write
    lock is in effect and read lock is requested (RT#7072)
  - 99dbfile.t, 99dbfilestore.t - added untie (for RT#6216)
  - Apache::Session::Lock::Semaphore will check for $Config{d_semget} and
    cygserver

1.82_02   2007-03-11 by Alexandr Ciornii
  - 99semaphore.t, 99flex.t will be skipped if $Config{d_semget}==undef,
    patch by Slaven Rezic
  - Removed redundant code in 99filelock.t
  - Removed unnecessary skip in 99nulllock.t
  - Added file 'Contributing.txt'
  - Apache::Session - mention CGI::Session
  - Request in Makefile.PL

1.82_01   2007-03-10 by Alexandr Ciornii
  - Applied part of patch of SREZIC (Slaven Rezic), RT#3670,
    more diagnostics on failing file operations (Apache::Session::Lock::File,
    Apache::Session::Store::DB_File, Apache::Session::Store::File)
  - RT#1251, ModUniqueId.pm, ModUsertrack.pm - small error in Carp usage,
    by Slaven Rezic
  - 99flex.t will be skipped on perls earlier than 5.8 (RT#16539)
  - Requires Storable (core from 5.7.3), as it is almost useless without it

1.82   2007-02-21 by Alexandr Ciornii
  - Slightly faster test skip
  - 99semaphore.t and 99flex.t check for cygserver on Cygwin
  - 99flex.t skips less tests in some cases
  - 99flex.t checks for Digest::MD5, MIME::Base64
  - Warning in doc of Apache::Session::Lock::Semaphore about cygserver
  - Requires File::Temp (core from 5.6.1), needed for tests and work
  - Changed order of use/chdir in tests (RT#16539, from Andrew Benham)

1.81_01   2007-02-12 by Alexandr Ciornii
  - My patch for Win32 and Cygwin (RT#18795)
  - More information in Makefile.PL
  - 99flex.t checks for Storable (RT#101)
  - 99semaphore.t does not die if IPC:: modules are not present

1.81   2006-05-23

  - Don't unconditionally try to require packages in Apache::Session::Flex
    (Dave Rolsky).

1.8    2005-10-06

  - Bug fix to stop death in Apache::Session::Lock::Semaphore.

1.70_01   2004-09-01

  - Casey West takes the pumpkin.
  - Complete rewrite of test suite to use Test::* modules.
  - Minor documentation tweaks.

---

Changes in 1.6 (2004-02-24):

Fixed file age test for clean().  The previous test was never true.

Applied patch to add TableName option to the DBI/MySQL store.  
From Corris Randall <corris AT line6.net>.  Tests added by jwb.

Applied patch from Oliver Maul <oliver.maul AT ixos.de> for better Sybase support,
without tests.

Informix support from Mike Langen <mike.langen AT tamedia.ch>.  Unfortunately no

Makefile.PL
use 5.005; #VERSION_FROM in Makefile.PL
#Symbol.pm requires 5.002
#5.004 almost 11 years ago, earliest version supported by perlver
#5.003_07 earliest version on CPAN

use ExtUtils::MakeMaker;

WriteMakefile1(
    NAME         => "Apache::Session",
    VERSION_FROM => "lib/Apache/Session.pm",
    MIN_PERL_VERSION => 5.005,
    META_MERGE => {
      resources=> {
        repository => 'http://github.com/chorny/Apache-Session',
      },
      keywords => ['CGI','session','web'],
    },
    'LICENSE'		=> 'perl',
    PREREQ_PM    => {
                     'File::Temp'      => 0, #core from 5.6.1
                     'Storable'        => '0.6', #Will be raised later
                     'constant'        => 0, #Available on CPAN now
#                     'MIME::Base64'    => 0, #new versions require 5.6
#                     'DB_File'         => 0, #test
                     'Digest::MD5'     => 0, #core from 5.8
                     'IO::File'        => 0, #core
                    },
    BUILD_REQUIRES => {
                     'Test::More'      => '0.47', #Build
                     'Test::Deep'      => '0.082', #Build
                     'Test::Exception' => '0.15', #Build
    },
    AUTHOR          => 'Casey West <casey[at]geeknest.com>',
    ABSTRACT        => 'A persistence framework for session data',
    'PL_FILES' => {},
);

sub WriteMakefile1 {  #Written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
  my %params=@_;
  my $eumm_version=$ExtUtils::MakeMaker::VERSION;
  $eumm_version=eval $eumm_version;
  die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
  die "License not specified" if not exists $params{LICENSE};
  if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
    delete $params{BUILD_REQUIRES};
  }
  delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
  delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
  delete $params{META_MERGE} if $eumm_version < 6.46;
  delete $params{META_ADD} if $eumm_version < 6.46;
  delete $params{LICENSE} if $eumm_version < 6.31;
  delete $params{AUTHOR} if $] < 5.005;
  delete $params{ABSTRACT_FROM} if $] < 5.005;
  delete $params{BINARY_LOCATION} if $] < 5.005;

  WriteMakefile(%params);
}

